s3pd/__init__.py

```diff
@@ -89,29 +98,42 @@ def create_client(signed=True):
             's3',
             config=botocore.config.Config(signature_version=botocore.UNSIGNED))
 
-def download_chunk(bucket, key, shmfileno, offset_first, offset_last, signed):
+def download_chunk(
+        bucket, key, shmfileno, offset_first, offset_last, signed, version=None):
     """Worker function to download a chunk of the file.
 
     :param bucket: Name of the S3 bucket.
     :param key: Path inside the bucket (without leading `/`)
     :param shmfileno: File descriptor for an opened destination file.
     :param offset_first: Start position of the chunk to download.
     :param offset_last: Last position of the chunk to download.
+    :param version: The file version to retrieve, or None
     :return: Nothing, the chunk is directly copied in the destination file.
     """
     client = create_client(signed)
+    args = {
+        'Bucket': bucket,
+        'Key': key,
+        'Range': 'bytes=%s-%s' % (offset_first, offset_last),
+        **({'VersionId': version} if version else {}),
+    }
     with shm_map(shmfileno, offset_first, offset_last - offset_first + 1) as shmmap:
-        chunk = client.get_object(
-            Bucket=bucket,
-            Key=key,
-            Range='bytes=%s-%s' % (offset_first, offset_last))['Body']
+        chunk = client.get_object(**args)['Body']
         chunk._raw_stream.readinto(shmmap)
 
-def resolve_link(bucket, key, client, depth=10):
+def resolve_link(bucket, key, client, depth=10, version=None):
+    """Resolve S3 links to target key.
+
+    :param bucket: Name of the S3 bucket.
+    :param key: Path inside the bucket (without leading `/`)
+    :param client: boto3 client to use when performing requests.
+    :param depth: Maximum number of link indirections before stopping.
+    :param version: The file version to retrieve, or None
+    """
     # Stop after too many link indirections
     assert depth > 0, 'Too many levels of link indirections'
 
-    filesize = get_filesize(client, bucket, key)
+    filesize = get_filesize(client, bucket, key, version=version)
```

Comment:
Shall we also pass the version id into the download_fileobj (line 144)?

Reply:
No.
The problem is that for links, `version` is very ambiguous. Is it the version of the link, or the version of the target? It could also be the version of any link in the resolving chain.
So I would prefer that when links are involved, versioning is just ignored.
I should even remove the version in `get_filesize` line 136, and add it to line 196.

Reply:
All good for me. 

My 2 cents is the version in the parameter list can be defined as version of the link, while the version of the target should be specified in the S3PDLINK, e.g. `#S3PDLINK#s3://test/test.txt&VersionId=<versionid>`.

Reply:
Agreed with @NewbiZ (at least for now, let's keep that part simple)